Convert plugin template to Circle CI 2.0

jdee opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Related to #10563. This updates the plugin template to use Circle CI 2.0, including test reporting.

### Description

Replaced circle.yml with .circleci/config.yml.

To enable test reporting, added `rspec_junit_formatter` as a dev dependency. RSpec now generates both doc-style output and a JUnit-style report.xml. The latter is used by CI for test reporting.

I played around with a number of options until I found all the right info in various places in the Circle docs to produce results like this:

![screen shot 2017-10-15 at 9 21 35 am](https://user-images.githubusercontent.com/121951/31586771-757f7eee-b18a-11e7-941b-aeec2131537f.png)

https://circleci.com/docs/2.0/configuration-reference/#store_test_results

Linked from that page ("classic CircleCI test metadata directory layout"):

https://circleci.com/docs/1.0/test-metadata/
https://circleci.com/docs/1.0/test-metadata/#rspec

jdee committed:
Convert plugin template to Circle CI 2.0

KrauseFx commented:
Awesome, thanks for your contribution on this @jdee, awesome idea. Let's see what @DanToml says about it :)

jdee commented:
One thought I had after submitting this: There's also Travis. That is, there's also a .travis.yml in the template. So devs have the option of not using what's here, and they may not use the `rspec_junit_formatter` gem. Of course, this is just a template, and they can modify whatever they like afterward. (They probably won't use both the .travis.yml and .circleci/config.yml, but they're both present, and one or both will probably be deleted.) It's nice for things to be set up well by default. And of course test reporting can also be set up for Travis. I think JUnit-style reporting is supported by most CI systems. I've used this gem with Jenkins before. So I don't think there's any harm including this dependency by default. But open to other opinions.

jdee committed:
Be sure to checksum Gemfile, not Gemfile.lock in Circle config

DanToml commented on fastlane/lib/fastlane/plugins/template/.circleci/config.yml:
```diff
@@ -0,0 +1,45 @@
+      - run:
+          name: install dependencies
+          command: |
+            bundle install --jobs=4 --retry=3 --path vendor/bundle
```
this should be `command: bundle check || bundle install --jobs=4 --retry=3 --path vendor/bundle`

DanToml commented on fastlane/lib/fastlane/plugins/template/.circleci/config.yml:
```diff
@@ -0,0 +1,45 @@
+
+      - save_cache:
+          paths:
+            - ./venv
```
I think we want the vendor folder here

DanToml commented on fastlane/lib/fastlane/plugins/template/.circleci/config.yml:
```diff
@@ -0,0 +1,45 @@
+      # run tests!
+      - run:
+          name: run tests
+          command: |
```
no need for multi line command here either

endocrimes reviewed (changes requested):
Good start!

jdee committed:
Address review comments

jdee committed:
Also bundle check before bundle install

jdee commented:
@dantoml Thanks. I think I addressed your concerns, particularly `bundle check` before `bundle install` and caching the vendor folder. (I think I was working from an old Circle template and never noticed that. All my gems thank you for catching it.)
